pkg/internal/controller/controller.go

```diff
@@ -144,6 +166,37 @@ func (c *Controller[request]) NeedLeaderElection() bool {
 	return *c.LeaderElected
 }
 
+// Warmup implements the manager.WarmupRunnable interface.
+func (c *Controller[request]) Warmup(ctx context.Context) error {
```

Comment:
I guess related to the error comment, having this return an `error` is very confusing because it will then get surfaced in the manager rather than in the controller when starting it - I'd prefer it if we had no return here and instead from the `sync.Once` either close `didWarmupFinishChan` on success or put the error in it.

That will reduce this method to just the sync.Once call if warmup is enabled

Reply:
> it will then get surfaced in the manager rather than in the controller when starting it

QQ: Are you saying here that if warmup failed with error, the controller's Start method should early exit out with that error? Is that the right way of getting the error to be surfaced in the controller when starting it?

Reply:
> having this return an error is very confusing

Also, to make warmup a runnable I need to have warmup return error. Are you saying that warmup should swallow the error, always return nil, then delay the error to be surfaced until the controller is started?

Reply:
> QQ: Are you saying here that if warmup failed with error, the controller's Start method should early exit out with that error? Is that the right way of getting the error to be surfaced in the controller when starting it?

Yes ideally. Call the sync.Once, block on the err chan, return err or continue.

> Also, to make warmup a runnable I need to have warmup return error. Are you saying that warmup should swallow the error, always return nil, then delay the error to be surfaced until the controller is started?

Either that or just no return and wrap it in the manager so it becomes a runnable

Reply:
Would this mean that `cm.runnables.Warmup.Start` returns without error even if it fails and the error would only be surfaced after the controller became leader?

Reply:
> Would this mean that cm.runnables.Warmup.Start returns without error even if it fails and the error would only be surfaced after the controller became leader?

Yes

Reply:
@godwinpang @sbueringer and I discussed this and came to the conclusion that it would be preferable to fail in `Warmup` already so we don't need to acquire the leader lease to error out. So I think what we want in the end is:
* Get rid of the `WaitForWarmupComplete`, we don't need it
* Both `Warmup()` and `Start()` will end up calling `startEventSources` which then calls the sync.Once and returns an error if any - Since we have a guarantee that the sync.Once only executes once, this is threadsafe
* Get rid of the `warmupResultChan` as we don't need it

Reply:
79a7b95

There are a few discussions so it is confusing, but I am pretty sure that reading between the lines we want to keep the warmup runnable returning error so I left it that way in the commit

Reply:
> but I am pretty sure that reading between the lines we want to keep the warmup runnable returning error so I left it that way in the commit

Yes! At a first glance the commit looks good. I'll do a full review tomorrow